### ISSUE TITLE
CE:FT ammo/armor/melee stat rebalance to 1.6+ standards

### ISSUE THREAD

**Issue by N7Huntsman** [AthenaSulisMinerva/CombatExtendedFastTrack#69]
Hey FT Team,

I'm just popping over from and looking to talk shop with someone regarding CE. There's a fairly major update to Combat Extended coming down the pipeline, and I wanted to discuss what all it is and how it will affect mod patches such as what you're working on. Please get in touch with me via Discord (N7Huntsman#4861), Steam, or the Ludeon Forums (N7Huntsman) when you have a few minutes to chat.

**Comment by AthenaSulisMinerva** [comment]
@N7Huntsman, I don't use Discord, Steam or forums. 

@sumghai and @Freedom187 do use them, but rather than having either of them play messenger and cause misunderstandings, I would appreciate it if we discuss things right here on GitHub instead.

**Comment by N7Huntsman** [comment]
As you like.

CE is pushing towards a 1.6 release in the immediate future--likely within the next week or two, based on what I've been told. Along with the introduction of some entirely new stuff (which, to my limited understanding of them, shouldn't directly impact patching mods), the armor penetration system is being reworked extensively. It's all written out in the Stat Sheets on the Combat Extended Github, but the short version is that projectiles and melee attacks will now have _two_ armor penetration values, which correspond roughly to a 'blunt' and 'sharp' penetration.

In addition to all the .xml work this necessitates directly, in many cases there's also been a fairly significant number of other adjustments related to it this; projectile speed/damage, ammo weight/bulk/recipe cost, etc. Needless to say, these are obviously all things that impact what you guys are doing.

With those things in mind, I wanted to see what your thoughts were on establishing a working relationship--or at least a line of communication--between FT and CE to cooperate in updating and creating compatibility patches. In the immediate, it would help put both versions on equal footing after the release hurdle, and in the longer term, it would help avoid duplication of work and ensure better consistency across patches.

Long cycle times for primary features of CE keeping compatibility patches from being folded into the main build is something that we've all recognized as an issue, and I think it's ultimately what you guys established FT to address. On the CE side, we've recently made a change wherein integrated patches will be pulled directly into the master branch after a once over, but my hope is that we can work together to help further improve the pipeline.

**Comment by AthenaSulisMinerva** [comment]
We've definitely heard rumors about the 1.6 armor penetration changes, and I've seen the new Stat Sheets on Google Docs. Thanks for coming by to confirm the rough timeline of the upcoming changes.

I'm interested in coming to a consensus regarding overlap in compatibility patches, but a subset of these patches will not be able to go into CE proper, for various reasons.

The CE:FT team currently collaborates using sumghai's copy of the stat spreadsheets for weapons, projectiles and animals. Most weapons use real-world ammo that already exists in the original CE, so that makes things easier for us in the event we have upstream changes like what you described.

**Comment by sumghai** [comment]
I'll switch my stat spreadsheets over to the new format once I have time.

At bare minimum, CE:FT and CE can work together on the stats spreadsheets for projectiles, melee weapons and animals/races.

As for ranged weapons, this is less critical since most firearms that CE:FT patches for aren't relevant to CE anyway (for instance, some really sci-fi mods with no real-world calibers).

Might be a good starting point if all the real-world ammo in the original CE repo were updated with the stats in xml, so we have an idea of what lines need to be changed.

**Comment by N7Huntsman** [comment]
Currently, all ammo, melee weapons, and animal melee attacks are in the process of being updated in the dev branch to the stats listed on the sheets. I'm working on the ammo types and hope to have everything PRed to the dev branch by the end of the day--melee weapons are updated, but other people are working on the animal melee, and my understanding is that it's pretty well in-hand. Some of them are already within the dev branch, so I'd take a glance through.

It's good that you bring up some of the more exotic projectile/ammo types like sci-fi weapons, because with only a limited amount of more exotic content within Combat Extended itself. With that in mind, if you guys play around with them a bit a develop an approach that seems to work reasonably well for projectiles that don't really fit the establish conventions CE's stat sheet uses to generate stats, I'd be interested to hear what you come up with.

Ultimately, I think the best outcome of collaborating would be both FT and CE coming out of this with a 1.6 build that is essentially as identical as possible in regards to mechanics, content balance, and compatibility patches. That way, both versions have the option to continue to work together like this in the future without having to working about mechanical differences getting in the way, as obviously, the two versions diverging to the point of incompatibility benefits no one,

**Comment by AthenaSulisMinerva** [comment]
I think it will be easier for us in the CE:FT if we simply wait until those stat rebalances (plus all the other bugfixes) in the CE dev branch get integrated back into the CE master branch. Then, we'll just make a PR and fix any conflicts/differences manually.

**Comment by N7Huntsman** [comment]
If that's what you prefer, though I'd be aware that I expect they'll be pushed to Master branch only on the day of release, and that may put FT behind the curve in terms to the functionality of their patches.

At any rate, some of the bigger ticket changes that I know will directly impact patching:
- Entirely new AP system for both ranged and melee attacks; Stats Sheets have everything laid out to make calculations for melee weapons and conventional projectiles (arrows, bullets, etc.), but I'm not necessarily sure how much exotic ammo types have been evaluated.
- Corresponding balance changes to damage as well.
- For ammo; new, formulaic approach to find values for work, ingredient cost, market value, etc.

I'm sure there'll be some other stuff that comes up before it's all said and done, but those are the major ones. 

Additionally, while we've both got other obligations in terms to patches to get up and running after the update, I wanted to offer the help of both myself and @DCheih , who has also done a lot of .xml work concerning updating and balancing stats for the 1.6 update. It wouldn't necessarily be completely unconditional, but we'd be happy to help.

**Comment by AthenaSulisMinerva** [comment]
> I'd be aware that I expect they'll be pushed to Master branch only on the day of release, and that may put FT behind the curve in terms to the functionality of their patches.

I'm sure the CE:FT team will be fine with that.

Speaking of these changes, how badly would they affect existing 1.5-era savegames? Many of us do have long-running saves that we would like to continue playing.

**Comment by N7Huntsman** [comment]
From what I've been told, the update itself won't necessitate a new save, so my guess would be the dictating factor will probably be just making sure all the compatibility patches you're using are up to date.

**Comment by sumghai** [comment]
I concur with Athena.

It shouldn't be hard for us to bring the remaining exotic ammo and projectiles up-to-date, and a couple of weeks' delay is a small price to pay for staying in sync with a major CE update.

**Comment by LockdownX7** [comment]
Regarding save compatibility, there will be a few console errors thrown when CE 1.6 is loaded into a 1.5 save. The game should gracefully recover from these for the most part, with no action needed by the player.

One exception applies for those using CE Guns, regarding loadouts containing the P7 or KRISS Vector, as those weapons were removed. These loadouts can no longer be edited, so they must be unassigned from any colonists that might be using them, then deleted.

**Comment by AthenaSulisMinerva** [comment]
Thanks guys.

Looking forward to bringing in the final changes when the time comes.

**Comment by sumghai** [comment]
I took a closer look at the current official CE Projectile Stats spreadsheet, and found that it was missing a lot of ammo definitions from stock CE. Conversely, I couldn't see the calculations for some of the newly-patched ammo in the CE 1.6 dev branch.

Therefore, I've made a copy of the spreadsheet, populated with all the ammo from both CE 1.6 dev branch and CE:FT, in the hopes that we can consolidate our calculation efforts (Yellow rows are CE:FT-exclusive ammo):

https://docs.google.com/spreadsheets/d/1-y3KCuTM_vJnxw-gQPl5KIiz7ySd7fw9P6Wx4zB5bZg/edit?usp=sharing

@AthenaSulisMinerva @Freedom187 @N7Huntsman @LockdownX7 - would you guys like editing permission to this spreadsheet?

**Comment by AthenaSulisMinerva** [comment]
Sure, sign me up. You have my email.

**Comment by AthenaSulisMinerva** [comment]
@sumghai - I know you've added myself and @Freedom187 to the projectile rebalance collab spreadsheet, but did you hear back from any of the CE team?

**Comment by sumghai** [comment]
Nothing from them at the moment, no.

By the way, I saw that you've brought over the commits from CE 1.6, and I had a play around with them. Other than some of our ammo no longer doing as much damage (which was expected), I haven't seen any errors.

Also, I think the reason the CE team haven't re-balanced all ammo types is because they (justifiably) would only focus on the ammo types that directly affect the CE Core patches and CE Guns. 

We're probably going to have to recalculate the stats for the rest of the ammotypes ourselves.

**Comment by sumghai** [comment]
@dchieh222, my understanding is that you've done some work with ammo stat updates. Would be interested in helping to fill in the rest of the joint CE / CE:FT spreadsheet?

**Comment by dchieh222** [comment]
Hello Robin,

I'm not aware of this joint spreadsheet and I work closely with
NoImageAvailable.  Could you let me know what is it?

Sincerely,
DChieh

**Comment by sumghai** [comment]
Hi DChieh,

Apologies for inviting you out of the blue - @N7Huntsman mentioned you earlier in the thread as someone who had experience with the new projectile stat balances.

Here's the link to the spreadsheet:

https://docs.google.com/spreadsheets/d/1-y3KCuTM_vJnxw-gQPl5KIiz7ySd7fw9P6Wx4zB5bZg/edit?usp=sharing

The structure is identical to the original CE projectile stat spreadsheet, except that I've gone through both CE and CE:FT to add all the unpatched/missing ammo defs, and ordered them alphabetically in the same order they show up on GitHub.

The yellow rows are ammo unique to CE:FT, which our team will assume full responsibility for (in addition to helping patch the remaining CE ammo). Once my access to Google Sheets is restored, I'll mark the rows that the CE team already patched in green.

**Comment by N7Huntsman** [comment]
You can give me edit access as well. I'm currently busy reviewing CE's integrated patches, but if nothing else, I can mark off the ones that have already been patched by CE.

**Comment by sumghai** [comment]
@N7Huntsman - In order to grant you edit access, I'll need your email address. To protect your privacy, send me a message on Discord and I'll sort you out.

I've just started marking off the patched ones myself, but having an extra pair of eyes to catch anything I've missed wouldn't hurt.

I've also noticed many of the "patched" rifle and pistol bullets actually have values of 0 for `armorPenetrationSharp` and `armorPenetrationBlunt`, and instead rely on `armorPenetrationBase`.

I'm assuming that this means you guys weren't able to find the appropriate gel & RHA penetration for most cartridges, and so are falling back to the old CE 1.5 penetration stats instead? And if so, if a modder has a new, unique ammo type, how would they calculate this old stat value in the event that their new ammo type also doesn't have publicly-available gel & RHA data?

**Comment by N7Huntsman** [comment]
Not all of the ammo projectiles within the current CE release have had their stats patched to 1.6--while most all of the common ammunition (NATO, Soviet-bloc, typical civilian) have been patched, there are a lot of one-off sets that haven't been patched yet. 

As I understand, `ammoPenetrationBase` is no longer the supported stat, and I don't know what exactly happens when you try to fire it in 1.6 (though I assume nothing all that great). In regards to ammo types that the RHA stats aren't readily available for or don't exist (such as fictional ammunition), it'll simply be a matter of making informed estimations or the author deciding how much armor they want their projectile to penetrate.

**Comment by sumghai** [comment]
Okay @N7Huntsman - you should have edit access now.

> Not all of the ammo projectiles within the current CE release have had their stats patched to 1.6--while most all of the common ammunition (NATO, Soviet-bloc, typical civilian) have been patched, there are a lot of one-off sets that haven't been patched yet.

I've started looking up values for many of these one-off sets - cartridge dimensions are relatively easy to look for via SAAMI, while bullet and case mass tend to be harder.

For case mass, do we need to account for (i.e. subtract) propellant weight as well?

> As I understand, `ammoPenetrationBase` is no longer the supported stat, and I don't know what exactly happens when you try to fire it in 1.6 (though I assume nothing all that great). The use of the new calculations is intended to remove as much of the guesswork as possible when patching real-world ammunition, but in regards to ammo types that the RHA stats can't be found for or don't exist (such as fictional ammunition), it'll simply be a matter of making informed estimations or the author deciding how much armor they want their projectile to penetrate.

Which probably explains why the 40x53mm grenades I added to the original CE currently does negligible damage (compared to the patched 40x46mm), among other things.

I've been making guesstimates for fictional ammo, such as using an online penetration calculator for railgun rounds. I'm generally more worried about missing values for real-life ammo.

**Comment by N7Huntsman** [comment]
The case mass should be the cartridge + propellant. It's part of the equation used to find the mass of the ammunition item.

**Comment by sumghai** [comment]
Much appreciated. I've added a note to the corresponding cell to clarify this for the rest of the team.

**Comment by sumghai** [comment]
@LockdownX7, I've been looking at the official [CE Projectile stats spreadsheet](https://docs.google.com/spreadsheets/d/1xnculiiWbFeGnEncwAPJ-DI6EoxwB7Bpl_8KGtWhXbU/edit#gid=1198674278) and some of your commits to the base CE repo, when I noticed some discrepancies in the `armorPenetrationSharp` and `armorPenetrationBlunt` stats for HE ammo types.

For example, the spreadsheet row for 40x46mm HE Grenades gives values of **None** for `armorPenetrationSharp` and `armorPenetrationBlunt`:

![Annotation 2019-11-06 202957](https://user-images.githubusercontent.com/4993380/68298427-4d68c500-00d4-11ea-835b-ccb6a41d20b1.png)

However, your commit in dfb88b0f08bdc9a0e5e79e5692474e4d26c97f78 specifies at least one non-zero armorPenetration value for that particular ammo:

![Annotation 2019-11-06 203228](https://user-images.githubusercontent.com/4993380/68298754-219a0f00-00d5-11ea-826d-d690cb185364.png)

How were these values calculated if the spreadsheet says they should be "None"?

**Comment by LockdownX7** [comment]
The spreadsheet is correct. Ignore the AP values from the commit you mentioned, as they're not even used. Explosives have their AP is derived as 10% of the before-armor damage that's applied to the target. Note that before-armor damage can vary due to falloff, based on distance from the explosion's center to its target(s), which in turn affects derived AP as well. In other words, all you need to set is how much damage the explosive should do on a direct hit, the code will calculate the rest.

In a similar vein, you don't need to worry about EMP AP either, as this damage type bypasses armor entirely.

**Comment by sumghai** [comment]
Thanks for the clarification, @LockdownX7.

So if I see a value of "none", do I omit those tags entirely? Or do I need to explicitly set them to zero?

**Comment by LockdownX7** [comment]
Omitting is fine.

**Comment by Theblessedwant** [comment]
I recently checked that the 1.6 update of CE added several patches for spartan Foundry, Vanilla expanded apparel and vanilla expanded armors, should the patches be merged with the upcoming FT version,?

**Comment by sumghai** [comment]
> I should clarify: when I say explosives don't require AP stats, I mean projectiles whose primary damageDef is Bomb. With that in mind, things like HEAT shells would still need explicit AP numbers, as their primary component is Bullet, with the Bomb set as secondary.

Makes sense! I assumed as much when I was patching some of the CE:FT missiles with HEAT warheads.

> I recently checked that the 1.6 update of CE added several patches for spartan Foundry, Vanilla expanded apparel and vanilla expanded armors, should the patches be merged with the upcoming FT version,?

They will be, yes - in fact, @AthenaSulisMinerva and I have already merged those patches from the upstream CE project into the master branch of CE:FT.

**Comment by Theblessedwant** [comment]
I have another question regarding balance, I checked the 5.56 and 5.45 ammo rounds and despite both being very similar irl, 5.45 for some reason is vastly superior over 5.56. For example, their sabot variants have the same base damage yet 5.45 has 21 armor penetration and 556 only has 14, this means that 545 can readily penetrate power armor while 556 cant even dream of such......
Additionally, the AP values for 545 has high armor penetration compared to 556 despite once again, 556 being a bigger round. This means there is absolutely no benefit for going with 5.56 weapons when with all variants they cant penetrate power armor or high tier armor vests. 5.45 does similar damage and can penetrate normal quality power armor too... This really needs addressing for the sake of balanace.

**Comment by Theblessedwant** [comment]
Furthermore, when comparing 762x39 and 762x51, Once again, 762x39 has a higher penetration as compared to 762x51 despite 762x51 being a much bigger round as compared to 762x39.  Even for the Sabot, 762x39 SABOT rounds will always penetrate even good power armor. while 762x51 SABOT rounds will struggle against even normal power armor.

**Comment by Freedom187** [comment]
Sounds like something is wrong there, should be as you say. (also yay github unborked my account finnaly!)

**Comment by sumghai** [comment]
@Freedom187 - Welcome back! You were sorely missed :)

@Theblessedwant - Thanks for bringing that to our attention - I'm currently still working through the different categories (as you can probably see by my color-coded rows in the spreadsheet), but I'll get to the rifle cartridges eventually.

Looking at the spreadsheet:
*  `armorPenetrationBlunt` correctly scales with bullet mass and muzzle velocity 
(e.g. `66.72` for `7.62x51mm NATO FMJ` vs. `41.080` for `7.62x39mm Soviet FMJ`), 

* `armorPenetrationSharp` depends on ballistic gel penetration data 
(which is _currently_ higher for the `7.62x39mm Soviet FMJ` than the `7.62x51mm NATO FMJ`)

Basically, I think it comes down to incorrect gel penetration values being put into the spreadsheet - there wasn't much documentation with the original version of the spreadsheet, but based on the values originally calculated by the CE Team for the 9x19mm Parabellum pistol cartridge, it looks like we're supposed to be using the 20% gelatin NATO protocol (as opposed to the 10% gelatin FBI protocol)

Regardless, I'll be going over every projectile at some point.

**Comment by LockdownX7** [comment]
Whenever possible, Sharp AP should be based on real world RHA penetration at 100 meters, gel data is just a fallback.

In light of this discussion, the CE team has reviewed the stats for the mentioned calibers, and increased sharp penetration for 5.56x45mm NATO. Current balance puts it at slightly more damage than the 5.45, in exchange for slightly less AP. The same relation applies to the 7.62 calibers.

Also keep in mind that there's more to combat than damage/AP. NATO caliber weapons have superior range and shot spread, making 5.56x45mm and 7.62x51mm better suited for long-range engagements, compared to their Soviet counterparts.

**Comment by AthenaSulisMinerva** [comment]
Keep up the good work, @sumghai - thanks for holding the fort while I deal with family issues this weekend.

**Comment by Theblessedwant** [comment]
Well i'd like to add that 762x51 is still doing way less penetration that 7.62x39. In data 7.62x51 does 21mm RHA penetration while 7.62x39 does 25mm RHA penetration. This makes no sense, based on data provided on bullet drop offs, 792x39 loses way more energy than 762x51 and furthermore has lesser starting energy. If we assume both are the same weight for simplicity, 762x51 travels much faster as compared to 762x51.
Utilizing simple kinetic energy calculations, we can find that, if both masses are the same, 7,62x51 imparts 3000~KJ of energy vs 7.62x39's 2000~ KJ of energy. 
Furthermore using newton's penetration impact depth approximation states that the impact changes with velocity. Greater velocity would imply greater penetration. 
What i suggest is switching the AP values for 762x51 and 762x39. It makes no sense for 762x39 to not deflect off power armor but 762x51 to deflect despite being a higher power round.

**Comment by sumghai** [comment]
Hi @LockdownX7, I've noticed that in the latest version of the Gun Stats spreadsheet, the **Ingame Stats** sheet no longer has the `recoilPattern` column, and that it is now possible for a firearm to sometimes have a `recoilAmount` of None

* Does this mean that when I come across None values, I can simply omit that stat? (Just like what you mentioned for certain stats in the Projectiles spreadsheet)

* Why is it that some two-handed and mounted weapons have no `recoilAmount`, while other weapons return a value? Is it because the calculations for the former determined that the actual number was negligible? Or some other criteria? (I couldn't quite decipher the formula used for that column)

**Comment by LockdownX7** [comment]
recoilAmount is the measure of accuracy loss for each shot fired in a single burst. Weapons that can only fire single-shot aren't affected by it, which is why the spreadsheest sets the stat to None.

You can omit the stat, it'll default to 0 as defined by the VerbPropertiesCE class.

**Comment by sumghai** [comment]
All standard and non-standard ammo and projectile calibers have been rebalanced.

Armor and melee tools/stat balances will be addressed on a case-by-case basis via submitted issues.
